draft/XXXX-robust-api/README.md

```diff
@@ -0,0 +1,146 @@
+
+### Additional Upload Endpoints
+
+`GET /api/workspaces/{workspace}/upload/{id}`
+
+This endpoint returns the Upload document with the corresponding ID.
+
+## Open Questions
+
+### Changes to data processing procedure
+
+The notion of having various file type "uploaders" will be removed. Instead, these will become data processors, which will take raw uploaded data (from S3), and run validation, table/graph creation, data insertion, etc. When downloading the finalized upload from S3, this processing should be done in a streaming manner, in order to limit the risk of running out of memory in the server. This will involve updating each data processor to handle streamed data, instead of the entire file. This could be done in the following manner:
+
+1. Read in data until a fully formed object can be parsed (a table row, for example).
+2. Run validation on this object
+3. If validation is passed, insert this object into the target location. If validation fails, save the validation error, and either continue processing more data, or halt data processing.
+4. Repeat steps 1-3 until all data has been processed
+5. If there are errors resulting from this processing, delete the partially formed tables/graphs, and return the validation errors to the user.
+
+These errors would likely be stored in the upload model (yet to be outlined), however, it may make sense to introduce another model that would track this status more specifically. Feedback is welcome here.
+
+
+### Tracking upload progress
+
+The current proposal requires that the client upload all parts, and then returns the ETags of these parts all at once when finalizing the upload. However, this poses a couple of possible pitfalls:
+
+1. There is no tracking of upload progress on the server
+2. If it's desired to allow for upload pausing and resuming (assuming the page is totally refreshed and all context is lost, or the computer is changed, etc.), there is no way to do this, as the initially acquired ETags are now lost
+
+This could be addressed by creating endpoints for finalizing an individual part upload, as well as the overall finalize call. In this case, the procedure would then require the client to call some endpoint with the following signature:
+
+```
+POST /api/workspaces/{workspace}/upload/{id}/finalizePart
+
+{
+  "partNumber": integer,
+  "ETag": string
+}
+```
+
+This would add to the `uploadPartTags` array on the upload document, which would allow for tracking the progress of this upload. After all file parts are uploaded, the upload finalize call is made as usual, but in this new approach, it would not contain any payload.
+
+
+Feedback on this suggested feature is appreciated.
+
+## Backwards Compatibility
+
+This RFC is not backwards compatible with the existing API, as the structure, upload procedure, etc. are all breaking changes.
```

Comment:
To fix the breakage, we'll need to update the main client, and change the multinetjs functions related to uploads. Any other changes?

Please record the list of what needs to be changed in this section.

Reply:
As far as I can see, the break in backwards compatibility are as follows:

- API
	- Uploads
	- Data Processing
- Client
	- Uploads/UI
- MultinetJS
	- Uploads

Am I missing anything?

Reply:
I think that's everything.